volatility3/framework/layers/arm.py

```diff
@@ -0,0 +1,475 @@
+        )
+        self._base_layer = self.config["memory_layer"]
+        # self._swap_layers = []  # TODO
+        self._page_map_offset = self.config["page_map_offset"]
+        self._tcr_el1_tnsz = self.config["tcr_el1_tnsz"]
+        self._page_size = self.config["page_size"]
+
+        # Context : TTB0 (user) or TTB1 (kernel)
+        self._virtual_addr_space = (
+            self._page_map_offset == self.config["page_map_offset_kernel"]
+        )
+        self._ttbs_tnsz = [self._tcr_el1_tnsz, self._tcr_el1_tnsz]
+
+        # [1], see D8.1.9, page 5818
+        self._ttbs_bitsizes = [64 - self._ttbs_tnsz[0], 64 - self._ttbs_tnsz[1]]
+        self._ttbs_granules = [self._page_size, self._page_size]
+
+        self._is_52bits = [
+            True if self._ttbs_tnsz[ttb] < 16 else False for ttb in range(2)
+        ]
+
+        # [1], see D8.2.7 to D8.2.9, starting at page 5828
+        self._granules_indexes = {
+            4: [(51, 48), (47, 39), (38, 30), (29, 21), (20, 12)],
+            16: [(51, 47), (46, 36), (35, 25), (24, 14)],
+            64: [(51, 42), (41, 29), (28, 16)],
+        }
+        self._ttb_lookups_descriptors = self._determine_ttbs_lookup_descriptors()
+
+        # [1], see D8.3, page 5852
+        self._descriptors_bits = [
+            (
+                49
+                if self._ttbs_granules[ttb] in [4, 16] and self._is_52bits[ttb]
+                else 47,
+                self._ttb_lookups_descriptors[ttb][-1][1],
+            )
+            for ttb in range(2)
+        ]
+
+        self._virtual_addr_range = self._get_virtual_addr_ranges()[
+            self._virtual_addr_space
+        ]
+
+        self._context_maxvirtaddr = self._ttbs_bitsizes[self._virtual_addr_space]
+        self._canonical_prefix = self._mask(
+            (1 << self._bits_per_register) - 1,
+            self._bits_per_register,
+            self._context_maxvirtaddr,
+        )
+
+        if AARCH64_DEBUGGING:
+            vollog.debug(f"Base layer : {self._base_layer}")
+            vollog.debug(
+                f"Virtual address space : {'kernel' if self._virtual_addr_space else 'user'}"
+            )
+            vollog.debug(
+                f"Virtual addresses spaces ranges : {[tuple([hex(y) for y in x]) for x in self._get_virtual_addr_ranges()]}"
+            )
+            vollog.debug(f"Pages sizes : {self._ttbs_granules}")
+            vollog.debug(f"TnSZ values : {self._ttbs_bitsizes}")
+            vollog.debug(f"Page map offset : {hex(self._page_map_offset)}")
+            vollog.debug(f"Descriptors mappings : {self._ttb_lookups_descriptors}")
+
+    def _determine_ttbs_lookup_descriptors(self) -> List[int]:
+        """Returns the bits to extract from a translation address (highs and lows)"""
+        ttb_lookups_descriptors = []
+
+        for ttb, ttb_granule in enumerate(self._ttbs_granules):
+            va_bit_size = self._ttbs_bitsizes[ttb]
```

Comment:
I see, I can add a direct function parameter instead of relying on the value prior existence. I can make a getter for the `granules_indexes` with a classmethod too :)